# Exploración del Global Copper Deposit Dataset

Primer acercamiento al dataset: carga, EDA y baseline con XGBoost para estimar ley de cobre.

## 1. Cargar datos

In [4]:
import pandas as pd

In [10]:
df = pd.read_excel("../data/external/Global-copper-deposit-dataset/Global_Copper_Deposit_Main.xlsx")
print(df.shape)
df.columns.tolist()
df.head()

(1487, 20)


,Mindat_id,Mindat_url,Deposit_name,Country,Latitude,Longitude,Deposit_type,Tonnage(Mt),Copper_grade(Cu; %),Molybdenum_grade(Mo; %),Gold_grade(Au; g/t),Silver_grade(Ag; g/t),Lead_grade(Pb; %),Zinc_grade(Zn; %),Nickel_grade(Ni; %),Cobalt_grade(Co; g/t),Max_age(Ma),Min_age(Ma),Mineral_assemblage,Chemical_Elements
0,641.0,Https://www.mindat.org/loc-641.html,Chuquicamata,Chile,-22.289444,-68.901111,Porphyry,21277.0,0.592,0.0400,0.013,5.0,NaN,NaN,NaN,NaN,35.07,31.0,"Albite,Aluminite,Alunite,Alunogen,Amarantite,A...",-Al-Na-Si-O-S-H-K-Fe-Ca-Cu-Cl-C-Ba-I-Mg-Ti-F-M...
1,2120.0,Https://www.mindat.org/loc-2120.html,Calabona,Italy,40.533056,8.334167,Porphyry,189.0,0.070,NaN,NaN,0.5,NaN,NaN,NaN,NaN,34.00,27.0,"Albite,Allophane,Alunite,Anhydrite,Azurite,Bor...",-Al-Na-Si-O-H-K-S-Ca-Cu-C-Fe-Mg-Ti-F-As-Sb-Pb-...
2,2355.0,Https://www.mindat.org/loc-2355.html,La Caridad,Mexico,30.320611,-109.564934,Porphyry,1800.0,0.452,0.0247,NaN,NaN,NaN,NaN,NaN,NaN,54.20,53.4,"Alunite,Anhydrite,Baryte,Bornite,Beryl,Chalcop...",-Al-O-K-S-H-Ca-Ba-Fe-Mg-Si-Ti-F-Cu-Be-As-Mo-Pb...
3,2596.0,Https://www.mindat.org/loc-2596.html,Moldova Noua,Romania,44.737778,21.702500,Porphyry,500.0,0.350,NaN,NaN,NaN,NaN,NaN,NaN,NaN,73.10,65.0,"Aikinite,Allophane,Analcime,Anhydrite,Anthophy...",-Bi-Cu-Pb-S-Al-Si-O-H-Na-Ca-Mg-C-Zn-As-Fe-Ti-K...
4,3326.0,Https://www.mindat.org/loc-3326.html,Christmas,USA,33.058333,-110.745833,Porphyry,48.1,1.100,NaN,0.010,1.0,NaN,NaN,NaN,NaN,64.00,63.0,"Actinolite,Andalusite,Andradite,Anhydrite,Anth...",-Ca-Fe-Mg-Si-O-H-Al-S-Cu-C-Ba-Ti-K-F-Zn-As-Pb-...


## 2. Estadísticas descriptivas

In [11]:
df.describe()

,Mindat_id,Latitude,Longitude,Tonnage(Mt),Copper_grade(Cu; %),Molybdenum_grade(Mo; %),Gold_grade(Au; g/t),Silver_grade(Ag; g/t),Lead_grade(Pb; %),Zinc_grade(Zn; %),Nickel_grade(Ni; %),Cobalt_grade(Co; g/t),Max_age(Ma),Min_age(Ma)
count,1483.000000,1483.000000,1483.000000,986.000000,989.000000,217.000000,546.000000,489.000000,190.000000,330.000000,39.000000,33.000000,1131.000000,1131.000000
mean,161031.552933,22.933827,5.768036,335.209189,1.280848,0.019057,1.115101,34.518181,1.333269,4.251246,0.672590,0.175818,634.209299,532.945217
std,100688.813688,30.153047,90.775315,1331.421283,1.358881,0.017925,3.648118,62.725241,1.617062,4.068063,0.871572,0.238836,881.961839,745.990849
min,125.000000,-54.829694,-160.670000,0.100000,0.029000,0.000500,0.001100,0.095000,0.010000,0.100000,0.110000,0.010000,0.000000,0.000000
25%,51441.500000,-0.564167,-77.944028,3.700000,0.430000,0.007000,0.180000,2.920000,0.272500,1.340000,0.315000,0.024000,57.695000,50.500000
50%,204650.000000,32.102920,13.102498,30.000000,0.800000,0.014000,0.400000,13.700000,0.800000,3.000000,0.400000,0.070000,270.000000,245.000000
75%,208392.500000,48.209314,92.575000,219.600000,1.640000,0.024000,1.000000,40.000000,1.700000,5.500000,0.610000,0.210000,738.500000,541.000000
max,455829.000000,73.680726,178.178333,21277.000000,13.000000,0.100000,70.000000,785.000000,8.800000,20.700000,4.380000,1.070000,4000.000000,3200.000000


## 3. Análisis de valores nulos y distribución de categorías

In [12]:
df.isnull().sum()

Mindat_id                     4
Mindat_url                    4
Deposit_name                  4
Country                       4
Latitude                      4
Longitude                     4
Deposit_type                  4
Tonnage(Mt)                 501
Copper_grade(Cu; %)         498
Molybdenum_grade(Mo; %)    1270
Gold_grade(Au; g/t)         941
Silver_grade(Ag; g/t)       998
Lead_grade(Pb; %)          1297
Zinc_grade(Zn; %)          1157
Nickel_grade(Ni; %)        1448
Cobalt_grade(Co; g/t)      1454
Max_age(Ma)                 356
Min_age(Ma)                 356
Mineral_assemblage           85
Chemical_Elements             5
dtype: int64

In [13]:
df["Deposit_type"].value_counts()

Deposit_type
Porphyry            640
VMS                 436
Sediment-Hosted     212
Magmatic sulfide    100
IOCG                 94
Porphyry Cu-Mo        1
Name: count, dtype: int64

In [15]:
 (df.notnull().sum() / len(df) * 100).round(1).sort_values(ascending=False)

Mindat_id                  99.7
Mindat_url                 99.7
Deposit_name               99.7
Country                    99.7
Latitude                   99.7
Longitude                  99.7
Deposit_type               99.7
Chemical_Elements          99.7
Mineral_assemblage         94.3
Max_age(Ma)                76.1
Min_age(Ma)                76.1
Copper_grade(Cu; %)        66.5
Tonnage(Mt)                66.3
Gold_grade(Au; g/t)        36.7
Silver_grade(Ag; g/t)      32.9
Zinc_grade(Zn; %)          22.2
Molybdenum_grade(Mo; %)    14.6
Lead_grade(Pb; %)          12.8
Nickel_grade(Ni; %)         2.6
Cobalt_grade(Co; g/t)       2.2
dtype: float64

## 4. Filtrar dataset modelable

Nos quedamos solo con registros que tienen `Copper_grade` (el target). De 1487 depósitos originales, 989 tienen ley de cobre reportada.

In [ ]:
df_cu = df.dropna(subset=["Copper_grade(Cu; %)"]).copy()
print(f"Registros con ley de cobre: {len(df_cu)}")
print()
print("Nulos en features principales:")
for col in ["Latitude", "Longitude", "Deposit_type", "Tonnage(Mt)", "Max_age(Ma)", "Min_age(Ma)"]:
    n = df_cu[col].isnull().sum()
    print(f"  {col}: {n} nulos ({100*n/len(df_cu):.1f}%)")

## 5. Visualizaciones

In [ ]:
import matplotlib.pyplot as plt

df_cu["Copper_grade(Cu; %)"].hist(bins=50, edgecolor="black")
plt.xlabel("Ley de cobre (%)")
plt.ylabel("Frecuencia")
plt.title("Distribución de Copper Grade")
plt.show()

In [ ]:
df_cu.boxplot(column="Copper_grade(Cu; %)", by="Deposit_type", figsize=(10, 5))
plt.suptitle("")
plt.title("Ley de cobre por tipo de depósito")
plt.ylabel("Cu (%)")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

## 6. Preparación para modelado

Definir features, target y encodear `Deposit_type` (variable categórica) a números.

In [ ]:
features = ["Latitude", "Longitude", "Tonnage(Mt)", "Max_age(Ma)", "Min_age(Ma)"]
target = "Copper_grade(Cu; %)"

In [ ]:
df_model = df_cu[features + [target, "Deposit_type"]].copy()
print(f"Registros: {len(df_model)}")
print(f"Nulos por columna:")
print(df_model.isnull().sum())

In [ ]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df_model["Deposit_type_encoded"] = le.fit_transform(df_model["Deposit_type"])

print(dict(zip(le.classes_, le.transform(le.classes_))))

In [36]:
from sklearn.model_selection import train_test_split

X = df_model[["Latitude", "Longitude", "Tonnage(Mt)", "Max_age(Ma)", "Min_age(Ma)", "Deposit_type_encoded"]]
y = df_model["Copper_grade(Cu; %)"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Train: {X_train.shape}")
print(f"Test:  {X_test.shape}")

# - test_size=0.2 — 80% para entrenar, 20% para evaluar
# - random_state=42 — para que sea reproducible

Train: (791, 6)
Test:  (198, 6)


## 7. Entrenamiento XGBoost

In [ ]:
from xgboost import XGBRegressor

model = XGBRegressor(
    n_estimators=300,
    max_depth=6,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42
)
model.fit(X_train, y_train)
print("Modelo entrenado!")

## 8. Evaluación del modelo

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

y_pred = model.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"RMSE: {rmse:.3f}")
print(f"MAE:  {mae:.3f}")
print(f"R²:   {r2:.3f}")

# - RMSE — error promedio en las mismas unidades (% Cu), penaliza errores grandes
# - MAE — error promedio absoluto, más intuitivo
# - R² — qué tan bien explica el modelo la variación (1.0 = perfecto, 0 = no explica nada)